diff --git a/src/xst.c b/src/xst.c
index 330ad44..7f60235 100644
--- a/src/xst.c
+++ b/src/xst.c
@@ -287,6 +287,7 @@ typedef struct {
 	int gm; /* geometry mask */
 	int tw, th; /* tty width and height */
 	int w, h; /* window width and height */
+        int hborderpx, vborderpx;
 	int ch; /* char height */
 	int cw; /* char width  */
 	int depth; /*  bit depth */
@@ -3578,8 +3579,10 @@ xhints(void)
 	sizeh->flags = PSize | PResizeInc | PBaseSize;
 	sizeh->height = xw.h;
 	sizeh->width = xw.w;
-	sizeh->height_inc = xw.ch;
-	sizeh->width_inc = xw.cw;
+//	sizeh->height_inc = xw.ch;
+//	sizeh->width_inc = xw.cw;
+        sizeh->height_inc = 1;
+        sizeh->width_inc = 1;
 	sizeh->base_height = 2 * borderpx;
 	sizeh->base_width = 2 * borderpx;
 	if (xw.isfixed) {
@@ -3886,8 +3889,8 @@ xinit(void)
 	xloadcols();
 
 	/* adjust fixed window geometry */
-	xw.w = 2 * borderpx + term.col * xw.cw;
-	xw.h = 2 * borderpx + term.row * xw.ch;
+	xw.w = 2 * xw.hborderpx + term.col * xw.cw;
+	xw.h = 2 * xw.vborderpx + term.row * xw.ch;
 	if (xw.gm & XNegative)
 		xw.l += DisplayWidth(xw.dpy, xw.scr) - xw.w - 2;
 	if (xw.gm & YNegative)
@@ -3997,7 +4000,7 @@ xinit(void)
 int
 xmakeglyphfontspecs(XftGlyphFontSpec *specs, const Glyph *glyphs, int len, int x, int y)
 {
-	float winx = borderpx + x * xw.cw, winy = borderpx + y * xw.ch, xp, yp;
+	float winx = xw.hborderpx + x * xw.cw, winy = xw.vborderpx + y * xw.ch, xp, yp;
 	ushort mode, prevmode = USHRT_MAX;
 	Font *font = &dc.font;
 	int frcflags = FRC_NORMAL;
@@ -4130,7 +4133,7 @@ void
 xdrawglyphfontspecs(const XftGlyphFontSpec *specs, Glyph base, int len, int x, int y)
 {
 	int charlen = len * ((base.mode & ATTR_WIDE) ? 2 : 1);
-	int winx = borderpx + x * xw.cw, winy = borderpx + y * xw.ch,
+	int winx = xw.hborderpx + x * xw.cw, winy = xw.vborderpx + y * xw.ch,
 	    width = charlen * xw.cw;
 	Color *fg, *bg, *temp, revfg, revbg, truefg, truebg;
 	XRenderColor colfg, colbg;
@@ -4224,7 +4227,7 @@ xdrawglyphfontspecs(const XftGlyphFontSpec *specs, Glyph base, int len, int x, i
 
 	/* Intelligent cleaning up of the borders. */
 	if (x == 0) {
-		xclear(0, (y == 0)? 0 : winy, borderpx,
+		xclear(0, (y == 0)? 0 : winy, xw.vborderpx,
 			winy + xw.ch + ((y >= term.row-1)? xw.h : 0));
 	}
 	if (x + charlen >= term.col) {
@@ -4232,7 +4235,7 @@ xdrawglyphfontspecs(const XftGlyphFontSpec *specs, Glyph base, int len, int x, i
 			((y >= term.row-1)? xw.h : (winy + xw.ch)));
 	}
 	if (y == 0)
-		xclear(winx, 0, winx + width, borderpx);
+		xclear(winx, 0, winx + width, xw.hborderpx);
 	if (y == term.row-1)
 		xclear(winx, winy + xw.ch, winx + width, xw.h);
 
@@ -4355,36 +4358,36 @@ xdrawcursor(void)
 		case 3: /* Blinking Underline */
 		case 4: /* Steady Underline */
 			XftDrawRect(xw.draw, &drawcol,
-					borderpx + curx * xw.cw,
-					borderpx + (term.c.y + 1) * xw.ch - \
+					xw.hborderpx + curx * xw.cw,
+					xw.vborderpx + (term.c.y + 1) * xw.ch - \
 						cursorthickness,
 					xw.cw, cursorthickness);
 			break;
 		case 5: /* Blinking bar */
 		case 6: /* Steady bar */
 			XftDrawRect(xw.draw, &drawcol,
-					borderpx + curx * xw.cw,
-					borderpx + term.c.y * xw.ch,
+					xw.hborderpx + curx * xw.cw,
+					xw.vborderpx + term.c.y * xw.ch,
 					cursorthickness, xw.ch);
 			break;
 		}
 	} else {
 	/* draw only cursor border: */
 		XftDrawRect(xw.draw, &drawcol,
-				borderpx + curx * xw.cw,
-				borderpx + term.c.y * xw.ch,
+				xw.hborderpx + curx * xw.cw,
+				xw.vborderpx + term.c.y * xw.ch,
 				xw.cw - 1, 1);
 		XftDrawRect(xw.draw, &drawcol,
-				borderpx + curx * xw.cw,
-				borderpx + term.c.y * xw.ch,
+				xw.hborderpx + curx * xw.cw,
+				xw.vborderpx + term.c.y * xw.ch,
 				1, xw.ch - 1);
 		XftDrawRect(xw.draw, &drawcol,
-				borderpx + (curx + 1) * xw.cw - 1,
-				borderpx + term.c.y * xw.ch,
+				xw.hborderpx + (curx + 1) * xw.cw - 1,
+				xw.vborderpx + term.c.y * xw.ch,
 				1, xw.ch - 1);
 		XftDrawRect(xw.draw, &drawcol,
-				borderpx + curx * xw.cw,
-				borderpx + (term.c.y + 1) * xw.ch - 1,
+				xw.hborderpx + curx * xw.cw,
+				xw.vborderpx + (term.c.y + 1) * xw.ch - 1,
 				xw.cw, 1);
 	}
 	if ((!strncmp(imstyle, IMSTYLE_OVERTHESPOT, 11)) &&
